libs/payments/qldb_types.go

```diff
@@ -0,0 +1,161 @@
+	Metadata     QLDBPaymentTransitionHistoryEntryMetadata     `ion:"metadata"`
+}
+
+// ValueHolder converts a QLDBPaymentTransitionHistoryEntry into a QLDB SDK ValueHolder.
+func (b *QLDBPaymentTransitionHistoryEntryBlockAddress) ValueHolder() *qldbTypes.ValueHolder {
+	stringValue := fmt.Sprintf("{strandId:\"%s\",sequenceNo:%d}", b.StrandID, b.SequenceNo)
+	return &qldbTypes.ValueHolder{
+		IonText: &stringValue,
+	}
+}
+
+func (e *QLDBPaymentTransitionHistoryEntry) toTransaction() (*AuthenticatedPaymentState, error) {
+	var txn AuthenticatedPaymentState
+	err := json.Unmarshal(e.Data.UnsafePaymentState, &txn)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"failed to unmarshal record data for conversion from qldbPaymentTransitionHistoryEntry to Transaction: %w",
+			err,
+		)
+	}
+	return &txn, nil
+}
+
+func AuthenticatedStateFromQLDBHistory(
+	ctx context.Context,
+	kmsClient wrappedKMSClient,
+	kmsSigningKeyID string,
+	stateHistory []QLDBPaymentTransitionHistoryEntry,
+	paymentState PaymentState,
+) (*AuthenticatedPaymentState, *QLDBPaymentTransitionHistoryEntry, error) {
+	latestHistoryEntry, err := validatePaymentStateHistory(ctx, stateHistory)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to validate history: %w", err)
+	}
+	signaturesAreValid, err := validatePaymentStateSignatures(
+		ctx,
+		kmsClient,
+		kmsSigningKeyID,
+		stateHistory,
+	)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to validate signatures: %w", err)
+	}
+	if latestHistoryEntry == nil || !signaturesAreValid {
+		return nil, nil, fmt.Errorf("state history failed validation: %v", paymentState)
+	}
+	authenticatedPaymentState, err := paymentState.ToStructuredUnsafePaymentState()
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get authenticated state from payment state: %w", err)
+	}
+	return authenticatedPaymentState, latestHistoryEntry, nil
+}
+
+// validatePaymentStateHistory returns whether a slice of entries representing the entire state
+// history for a given id include exclusively valid state transitions.
+func validatePaymentStateHistory(
+	ctx context.Context,
+	transactionHistory []QLDBPaymentTransitionHistoryEntry,
+) (*QLDBPaymentTransitionHistoryEntry, error) {
+	var (
+		reason                    error
+		err                       error
+		unmarshaledTransactionSet []AuthenticatedPaymentState
+	)
+	// Unmarshal the transactions in advance so that we don't have to do it multiple
+	// times per transaction in the next loop.
+	for _, marshaledTransaction := range transactionHistory {
+		var transaction AuthenticatedPaymentState
+		err = json.Unmarshal(marshaledTransaction.Data.UnsafePaymentState, &transaction)
+		if err != nil {
+			return nil, fmt.Errorf("failed to unmarshal transaction data: %w", err)
+		}
+		unmarshaledTransactionSet = append(unmarshaledTransactionSet, transaction)
+	}
+	for i, transaction := range unmarshaledTransactionSet {
+		// Transitions must always start at 0
```

Comment:
question: does this mean that the first transaction should always be prepared? The comment and `if` statement are a little unclear.

question: do we need to check `i == 0` on every iteration?

suggestion (non-blocking); would this make it clearer

```
... necessary len checks

if unmarshaledTransactionSet[0].Status != Prepared {
	return nil, &InvalidTransitionState{}
}

for i := 1; i < len(unmarshaledTransactionSet); i++ {
	from := unmarshaledTransactionSet[i-1]
	to := unmarshaledTransactionSet[i]

	if !from.NextStateValid(to.Status) {
		return nil, &InvalidTransitionState{
			From: string(from.Status),
			To:   string(to.Status),
		}
	}
}
```